### ISSUE TITLE
0002: All builds use a caching proxy for Maven and Python packages

### ISSUE THREAD

**Issue by mitchhentges** [mozilla-releng/releng-rfcs#4]
# Summary

Cache our dependencies in a "caching proxy" rather than putting them in docker images

## Motivation

TL;DR: Improve performance and maintainability

We used to run into issues with putting too much load on official Maven and Python package repositories - so much that they would rate limit us. We created [a bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1493626) to represent this issue which suggested using a package caching proxy to resolve the problems.

In the meantime, we've been caching dependencies in our docker images. This has a few downsides:
* Whenever dependencies change, we need to build a new Docker image, otherwise we're at risk of over-hammering package repositories
* Our images are very large, which is increasing image-build time and causing a longer task run time (need to download the massive file)
* Complexity: if there's an issue with a package (out-of-date, wrong version, corrupt), it's harder to track down (is the docker image correct? Has it been rebuilt? Did it download correctly? Did it build correctly?)
* Intermittent failures: these are more common with this Docker solution, because if we forget to release a new Docker image after updating dependencies, it won't _immediately_ fail since we'll only be requesting from the package repository for these new dependencies. However, as we continue changing dependencies, the chances of failure will increase. When we start seeing failures pop up "randomly", it'll be difficult and annoying to track down.

The relative downsides for running a package proxy are significantly lesser:
* Need to run additional machines (though they don't need to be very CPU-heavy)
* Complexity: just like with the docker image, if there's issues with a package, there's increased work required in tracking down the issue. Fortunately, this solution is a lighter abstraction, so detecting issues is arguably easier

[St3fan did a test of using a package proxy](https://mozilla.slack.com/archives/CA4JKKLSG/p1545158079260400) and it sounds like he found success - he didn't elaborate on any issues he found (not to say there won't be any, but it implies that there weren't immediate blockers)

# Details:

We probably want a caching proxy that operates at the package level. It should work something like this:

1. Task `1` requests for package `p:v1` to be downloaded:
```
task                caching proxy               repository
       -->  p:v1?
                    don't have it!
                                      -->  p:v1?
                                      
                                                  has p:v1 (and many other packages)
                                      
                                          |----|
                                      <-- |p:v1|
                                          |----|
                    has p:v1
           |----|
       <-- |p:v1|
           |----|
           
has p:v1
```

2. Now, task `2` requests for package `p:v1` to be downloaded (but the caching proxy has now fetched it, so doesn't have to ask the original repo anymore):
```
task                caching proxy               repository
       -->  p:v1?
                    has p:v1
           |----|
       <-- |p:v1|
           |----|

has p:v1
```

3. < ... many builds, added dependencies, etc ...>
4. Task `n` is run, and changes one of the dependencies:
```
task                caching proxy               repository
       -->  p:v1?
            q:v1?
            r:v2?
            s:v1?
            t:v1?

                    has p:v1
                        q:v1
                        r:v2
                                      -->  s:v1?
                                           t:v1?


                                                  has packages

                                          |----|
                                      <-- |s:v1|
                                          |----|

                                          |----|
                                          |t:v1|
                                          |----|

                    has p:v1
                        q:v1
                        r:v2
                        s:v1
                        t:v1
           |----|
       <-- |p:v1|
           |----|

           |----|
           |q:v1|
           |----|

           |----|
           |r:v2|
           |----|

           |----|
           |s:v1|
           |----|

           |----|
           |t:v1|
           |----|

has p:v1
    q:v1
    r:v2
    s:v1
    t:v1
```

There exists software already that can do this for both Maven/Gradle (St3fan used Sonatype) and Python ([`devpi-server`](https://pypi.org/project/devpi-server/)?). I'll leave the software-selection up to the implementation

# Open Questions:

* We're already leaning on Docker images having dependencies cached, right? If yes, as part of this RFC, that part should be rolled back so our images don't contain unused caches
* What needs to be accomplished for setting this up?
    * Have a machine provisioned to act as our caching proxy (ask the ops team)
        * Is it ok to have only one machine for both Python caching and Maven caching?
        * Since it's just network and disk traffic, I'm assuming that the load will be small enough that one machine will be able to handle this easily - do we have data on how much traffic we'll be using? Perhaps we can start with one machine, then bump up as we roll out tasks pointing to this caching proxy?
    * Should releng set up the caching software on this machine and maintain it?
        * How do we want to handle updates to the machine itself (caching proxy software, linux updates that require restarts)?
* Should we have builds fall back to the main repositories if our caching proxy is down? (IMHO, if we set up a zero-downtime proxy, It'd be less maintenance/variability if we just always depended on our caching proxy, so we see immediately if it goes down)

**Comment by escapewindow** [comment]
I don't know the specifics of this proposal. However, we've dealt with versions of proxy caches before. I believe we should adhere to the following:

- Strict separation of level 1 and level 3 tasks and caches. If someone is able to "poison" a cache via Try, an unreviewed PR, or an unofficial repo, it's problematic if we then craft a release using that cache.
  - Similarly, the level 3 cache should be somehow restricted to official release graphs.
- Cross-region caches for both network costs and robustness. No new SPOF.
- It's possible that Mobile is small enough in scale that downloading a module on request works just fine. In Gecko land, we may have dozens or hundreds of tasks all requesting the same un-downloaded dependency at the same time. In this case, how the proxy behaves is extremely important. Serving a partially downloaded dep is bad. Downloading dozens or hundreds of copies of the same dep concurrently is bad. Overwriting another download's file with a new file, causing the previous download to fail, is bad. OOMing is bad. and so on.
- We need metrics, a way to administrate, docs, and ownership.

Recommended:
- It may be wise, when possible, to include shas as well as pinned versions. Best practices says that once an upstream dependency is released, the maintainers should never remove and re-release it. We've seen examples of the opposite in the wild. This is much worse if and when we add Try/level 1 testing PRs that download arbitrary versioned deps; we may get test version collisions.
- It may be desirable to have a way to clear the cache At the very least, tools to identify bad-cache workers and clearing them out. Even if everything else goes well, sometimes the cache can partially download a dependency without erroring out and then everything afterwards is broken. And there's the chance of other cache issues.

**Comment by escapewindow** [comment]
Taskcluster already has code for cloud-mirror that has a robust cache of files that gets copied around  different AWS regions; we may be well-served in basing our code off that. And still, I believe we still see intermittent issues with cloud-mirror that cause us to download broken packages over and over until someone can squash the bad file(s).

**Comment by mitchhentges** [comment]
* Regarding poisoning the cache: all of our caches for different levels should pull from the same sources (E.g.: pypi). Since you can't "push" to a cache, but only request that a cache download from the official source, the risk of poisoning depends on the caching software itself failing - at which point, if a level 1 cache can be poisoned, what stops a level 3 cache from being poisoned?
* Multiple caches sounds :+1:
* How multiple downloads happen at once is definitely important to handle correctly.
* Metrics/administration/docs/ownership :+1:

The concerns being listed aren't specific to us: every package caching proxy should:
* Not be poisonable
* Handle parallel downloads intelligently
* Include SHAs

I strongly recommend that we look for an existing solution before considering building this ourselves.

**Comment by JohanLorenzo** [comment]
> if a level 1 cache can be poisoned, what stops a level 3 cache from being poisoned?

I agree with the scenario you mentioned, the risk is limited. Having them split, prevents this (unlikely but plausible) scenario:
0. Our dependencies aren't pinned. For instance, the dependency manifest says to always pull the latest. 
1. An attacker managed to poison a lightly secured dependency.
2. Someone (not necessarily the attacker) opens a PR, the cache now delivers the poisoned dependency. 
3. The dependency is removed from the official repositories.
4. Even though the PR never landed, the cache delivers what it thinks is the latest dep: the poisoned one. Next release is built with that dep included. 

> I strongly recommend that we look for an existing solution before considering building this ourselves.

I agree. We can also look into TC's cache. It supports levels 1,2, 3. A discussion started in https://github.com/mozilla-mobile/android-components/issues/1462

**Comment by escapewindow** [comment]
Older pips sometimes guessed at filename -> package name. Newer pip has addressed this, but I'm not sure if that's comprehensive. So, for instance, if pip has the issue of sometimes ignoring stuff after dashes, someone could upload a `scriptworker-fake-$version.tar.gz` and then someone trying to install scriptworker might get `scriptworker-fake` instead.

Newer pip should [mostly?] address. Pinning versions helps. Pinning shas everywhere makes this problem [completely?] go away, but adds its own concerns and need for tooling. And on try or unreviewed PRs, the requirement for pinning shas might be bypassed.

If we can somehow guarantee that cache poisoning cannot ever affect release tasks, then a shared cache works. If there's some grey area or some known issues, I think it's straightforward enough to add a 2nd cache. I'm not sure how you prevent using the level 3 cache from a level 1 task, if all the checks are in-repo code; someone could modify that in their try push / unreviewed PR. If we use taskcluster caches, those are separated by `workerType`, so the user would have to get scopes to schedule their task on a level 3 worker.

**Comment by mitchhentges** [comment]
> 1. An attacker managed to poison a lightly secured dependency.

> someone could upload a  ...

One thing to keep in mind is that this will behave identically to our current setup with Docker - when we populate the cache, we download the packages from the same official repositories.
Of course, there is a risk of attackers somehow poisoning the cache/docker image - with Docker, we're trusting that our images (that aren't verified by digest!) aren't maliciously modified. In the same way, with package-caching software, we have a risk of the packages being maliciously modified (e.g.: if the machine is compromised). I'm not sure that this risk would change :thinking: 

Related: regardless of our solution, I think that pinning versions+shas is fantastic, and probably necessary :smile:

**Comment by mitchhentges** [comment]
Also, regarding TC caching: my understanding is that this caches at the file-system level?
Keep in mind a couple downsides:
* Like docker, this caches intermediary files that we don't care about (lockfiles, `gradle`/`pip` files created from processing the data (denormalization?), extra metadata, etc)
* The concern of purging old files: if a dependency is removed, how do you remove the now-unnecessary tree of unused files? We could auto-purge every `$amountOfTime`, but then our build times will have spikes each time the cache is purged

**Comment by JohanLorenzo** [comment]
> One thing to keep in mind is that this will behave identically to our current setup with Docker - when we populate the cache, we download the packages from the same official repositories.

I agree this can happen. However, the likeliness changes in the Docker scenario => Instead of getting the dep at every single event that occurs on a PR, we restrict the window to the times we build a new docker image, which is every other week or so. 


> Docker, we're trusting that our images (that aren't verified by digest!)

Excellent point. We can switch to digests quite easily. We can handle once https://github.com/mozilla-mobile/android-components/pull/1729 is done. Would you have the bandwidth to make the change @mitchhentges ? I can review.


> In the same way, with package-caching software, we have a risk of the packages being maliciously modified (e.g.: if the machine is compromised)

I agree. We can mitigate the risk by lowering the likeliness of the machine being compromised. This can be achieved by making it less open to external requests (e.g.: pull requests). That's why having 2 caches is a bonus: we have a restricted instance for releases and an open one for people to work efficiently on PRs. 


> Like docker, this caches intermediary files that we don't care about (lockfiles, gradle/pip files created from processing the data (denormalization?), extra metadata, etc)

True. We may be able to exclude them for the cache. This sounds like the type of issues a Firefox builds would have. 


> The concern of purging old files: if a dependency is removed, how do you remove the now-unnecessary tree of unused files?

Excellent question. I confess I have no idea. Maybe @ncalexan has some insights.

**Comment by mitchhentges** [comment]
Thanks :)

One note:

> I agree this can happen. However, the likeliness changes in the Docker scenario => Instead of getting the dep at every single event that occurs on a PR, we restrict the window to the times we build a new docker image, which is every other week or so.

When you say "getting the dep at every single event", are you referring to getting the dep from the official repositories? With a package proxy, we only download each package **only once** from the official repositories. (If you're referring to downloading from the caching proxy, we _would_ download the package once-per-event from that, but this isn't a problem because it's functionally the same or less to downloading the docker image once-per-event).

**Comment by ncalexan** [comment]
> > The concern of purging old files: if a dependency is removed, how do you remove the now-unnecessary tree of unused files?
> 
> Excellent question. I confess I have no idea. Maybe @ncalexan has some insights.

I see two times when purging is relevant:

1. the Maven caching repository itself.  Generally, the contract is that Maven repositories are "append only".  Many things will break in subtle ways if that contract is violated, although they tend not to actually matter.  I wouldn't be concerned about purging ancient versions from a Maven caching repository.

2. the local `.gradle`/`.m2` cache on build machines themselves.  It makes the most sense to me to use TC volumes for these caches.  Such volumes come and go, and that's actually a pretty reasonable form of LRU cache right there :)  It does push management of the cache to the individual builds, leading to spikes both when the build is fresh (i.e., when the volume is empty) and when the build is stale (i.e., when the volume is full, dependencies roll forward, and significant clean-up must be done).  Neither of those situations worry me.  When the build is fresh, our Maven caching repo lets us populate it efficiently.  When the build is stale, I trust Gradle to manage it's LRU reasonably efficiently.  I wouldn't invest much here until I was confident spikes to clean-up were expensive in the wild.

**Comment by mitchhentges** [comment]
@ncalexan if we were to use a caching proxy, what value would a TC volume `.gradle` cache add on top?
My understanding is that, with a TC volume, it will move the cache over the network and propagate it to each new build machine.
Assuming that the caching proxy _also_ exists on the same network (e.g.: it's not _faster_ when the TC volume is used), I missing the value of TC volume for this use-case :thinking:

**Comment by ncalexan** [comment]
> @ncalexan if we were to use a caching proxy, what value would a TC volume `.gradle` cache add on top?
> My understanding is that, with a TC volume, it will move the cache over the network and propagate it to each new build machine.
> Assuming that the caching proxy _also_ exists on the same network (e.g.: it's not _faster_ when the TC volume is used), I missing the value of TC volume for this use-case 🤔

It may provide none.  I, very sketchily, thought that these TC volumes were provided by separate instances, not transferred over the network but just "attached" and traded around.  Anyway, you are expert in these things, not me -- if there's no benefit to `.gradle`, don't cache it!

I will note that other things still want these types of TC volume caches: in particular, we `sccache` locally in Application Services.  I guess we could `sccache` globally (i.e., back with cloud storage) for a-s as well...

**Comment by JohanLorenzo** [comment]
> When you say "getting the dep at every single event", are you referring to getting the dep from the official repositories? With a package proxy, we only download each package only once from the official repositories.
I'm sorry for the confusion. I mean, in the Docker image scenario, we refresh the cache (stored in the image) every time we rebuild the image. In the single proxy scenario, the cache gets refreshed every time something hits it. It can be a release, a build from the master branch, an unreviewed PR. The latter is concerning if it's shared with release builds. That's why I think the likelihood is reduced in the Docker scenario. 

> Assuming that the caching proxy also exists on the same network [...] I missing the value of TC volume for this use-case
I agree, if the proxy is on the same network, performance should be approximately the same. I see 2 values in favor of a TC volume:
* the feature exists, it's known and maintained by the TC team. Releng and probably RelOps don't have to spin up X instances of this proxy, then configure and maintain it.
* it's not specific to a single type of package, so we don't have to redo the same process if we want to support python packages, for instance. 

That said, TC volumes may be too complex and hard to work with. I haven't worked with them/. I think we should weight in the cost of both solutions. What do you think?

**Comment by mitchhentges** [comment]
Ah, I think I see difference in expectation regarding "getting the dep at every single event".
With a package proxy, one of the requirements would be that if:
1. A build requests package "asyncio" at version "3.0.0"
2. And the package proxy has downloaded `asyncio@3.0.0` before
3. Then the package proxy **will not** request `asyncio` from the official repositories again.

I tried to represent this in the [initial comment](https://github.com/mozilla-releng/releng-rfcs/issues/4#issue-397138316) in the 2nd example in "Details". It was a noisy comment, sorry if it wasn't clear :)

TL;DR: The cache won't be refreshed each time it gets hit. Since `$package@$version` always has the same content, there's no need to re-query the official repositories.
The only case that that would be necessary would be if our build requested the _latest_ version of a package, in which case we'd need to hit the official repositories every time (this should be an edge case and should be actively avoided)

-----

Regarding TC volumes, I'd like to document what was discussed in [slack](https://mozilla.slack.com/archives/CDKB7QPCL/p1547746703064200).
TC volumes apply to a single instance. They:
* aren't sent over the network.
* are tied to instances that are recycled every 24-48 hours, at which point the TC cache needs to be rebuilt
* aren't fully utilized on an instance if there are multiple jobs in parallel on that instance

So, using TC volumes would reduce traffic between our package proxy and our build workers. It would be valuable to use TC volumes, but they would work in tandem with an additional caching solution - TC volumes won't be sufficient on their own

**Comment by mitchhentges** [comment]
I think there's enoguh discussion here that I can officially RFC this and put it in the "proposal" phase

**Comment by ncalexan** [comment]
> So, using TC volumes would reduce traffic between our package proxy and our build workers. It would be valuable to use TC volumes, but they would work in tandem with an additional caching solution - TC volumes won't be sufficient on their own
 
This point is minor, but: no, it's not true that TC volumes are sufficient.  a-s had working builds with:
- no package proxy and no TC volumes
- no package proxy and TC volumes
and presumably could have working builds with
- package proxy and no TC volumes
- package proxy and TC volumes
a-c looks different -- it has more dependencies -- but we haven't demonstrated that TC volumes alone won't address the intermittent failures.  In fact, I believe they will; but I don't feel strongly enough to try to ensure that we try them alone before doing other, more significant things.

**Comment by mitchhentges** [comment]
I think that there's two concerns here:
1. Can we avoid intermittent failures due to overloading the official repositories
2. Can we improve build performance

Fortunately, performance should be correlated with how much we hit the official repositories. Assuming that "`tc_volume_when_populated` faster than `download from caching proxy` faster than `download from official repository`":

| Using package proxy | Using TC volume | Assumed performance | Why |
|:-----:|:-----:|:-----:|-----|
| no | no | very slow | Always hitting real package repositories
| no | yes | slow | TC cache won't always be populated
| yes | no | fast | Relatively fast, but always fetching existing packages over (local) network
| yes | yes | very fast | Either uses local cache (really fast), or falls back to fetching existing packages over (local) network

So, to answer:

> Will the workers download asyncio==3.0.0 from the proxy every run?

If we use TC volumes, then `asyncio==3.0.0` will only be downloaded from the proxy if it isn't in the TC cache (so: sometimes :smile:). Of course, `asyncio==3.0.0` would only have been downloaded from the official repositories just once.
